classes/opensesameapi.php

```diff
@@ -255,69 +273,86 @@ public function determineurl(object &$paging): bool {
      * @throws \stored_file_creation_exception
      */
-    public function get_open_sesame_course_list(string $token, string $url): bool {
+    public function get_open_sesame_course_list(): bool {
         global $DB;
+        mtrace('Getting Opensesame Course List.');
         // Integrator issues request with access token.
-        $this->setHeader(['content_type: application/json', sprintf('Authorization: Bearer %s', $token)]);
-        $response = $this->get($url);
-        $statuscode = $this->get_http_code();
-        $dcoded = json_decode($response);
-
-        if ($statuscode === 400) {
-            mtrace('OpenSesame Course list Statuscode: ' . $statuscode);
-            throw new \moodle_exception('statuscode400', 'tool_opensesame');
-        }
-        if ($statuscode === 200) {
-            mtrace('OpenSesame Course list Statuscode: ' . $statuscode);
-            $paging = $dcoded->paging;
-            $data = $dcoded->data;
-
-            foreach ($data as $osrecord) {
-
-                $this->create_oscategories($osrecord);
-                $keyexist =
-                        $DB->record_exists('tool_opensesame', ['idopensesame' => $osrecord->id]);
-                $osdataobject = new \stdClass();
-                $osdataobject->idopensesame = $osrecord->id;
-                $osdataobject->provider = 'OpenSesame';
-                $osdataobject->active = $osrecord->active;
-                $osdataobject->title = $osrecord->title;
-                $osdataobject->descriptiontext =
-                        $osrecord->descriptionHtml ? $osrecord->descriptionText : $osrecord->descriptionHtml;
-                $osdataobject->thumbnailurl = $osrecord->thumbnailUrl;
-                $osdataobject->duration = $osrecord->duration;
-                $osdataobject->languages = $osrecord->languages[0];
-                $osdataobject->oscategories = implode(', ', $osrecord->categories);
-                $osdataobject->publishername = $osrecord->publisherName;
-                $osdataobject->packagedownloadurl = $osrecord->packageDownloadUrl;
-                $osdataobject->aicclaunchurl = $osrecord->aiccLaunchUrl;
-
-                if ($keyexist !== true) {
-                    $returnid = $DB->insert_record('tool_opensesame', $osdataobject);
-                    mtrace('inserting Open-Sesame course ' . $osrecord->title . ' metadata. id: ' . $returnid);
+        // Reset headers for a new request.
+        $this->resetHeader();
+        $this->setHeader(['content_type: application/json', sprintf('Authorization: Bearer %s', $this->access_token)]);
+        $maxattempts = 4;
+        $retrycount = 0;
+        while ($retrycount < $maxattempts) {
+            mtrace('This courserequesturls '. $this->courserequesturl);
+            $response = $this->get($this->courserequesturl);
+            $decoded = json_decode($response);
+
+            $statuscode = $this->get_http_code();
+
+            if ($statuscode === 400) {
```

Comment:
If we are going to retry 400 errors, why not other error codes?

Reply:
I'm not sure why we would want to retry on a 400 even since that means "bad request."  If we get back a 400 (bad request) but we don't change anything about the request, we should not expect to get anything other than a 400 back again.  The list of status codes to retry on as a general rule would be:
408 Request Timeout
425 Too Early
429 Too Many Requests
500 Internal Server Error
502 Bad Gateway
503 Service Unavailable
504 Gateway Timeout

Reply:
The updated pull request addresses these codes. As for the status 400 this will result in a Moodle_exception stoping the scheduled task completely. The code has been improved as to address the previous issue that was causing the status 400 and if this was to occur in the future it would signify an issue a developer would need to address. Therefore you would see opensesamesync scheduled task has failed along with the error code.  Moodle will reattempt the scheduled task the next time it is scheduled to run in the system and will continue to do so until the issue is resolved. Essentially the migration of courses would not be possible and it would mean something has changed in the api request. The new code address this issue.

Reply:
Looks good!  Thank you